osu.Game/Screens/Play/Player.cs

```diff
@@ -181,6 +274,21 @@ protected override bool OnExiting(GameMode next)
             return base.OnExiting(next);
         }
 
+        protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
+        {
+            switch (args.Key)
+            {
+                case Key.Escape:
+                    if (!IsPaused && clockWasStarted) // For if the user presses escape quickly when entering the map

```

Comment:
you should be able to override gamemode's OnExiting and add this pause logic there. return true in the override in the cases you don't want to actually exit. this way you won't even need to specify the escape key (and it will likely work in more cases in the future as a result).